xray/src/bin/merge_xray_quadtrees.rs

```diff
@@ -20,8 +27,179 @@ struct CommandlineArguments {
+    directories
+        .iter()
+        .map(|directory| read_metadata_from_directory(&directory))
+        .flatten()
+        .collect()
+}
+
+fn get_root_node(meta: &proto::Meta) -> Option<NodeId> {
+    meta.get_nodes()
+        .iter()
+        .map(NodeId::from)
+        .min_by_key(|node| node.level())
+}
+
+fn get_root_nodes(meta: &Vec<proto::Meta>) -> Vec<Option<NodeId>> {
+    meta.iter().map(get_root_node).collect()
+}
+
+struct Metadata {
+    root_nodes: FnvHashSet<NodeId>,
+    nodes: FnvHashSet<NodeId>,
+    level: u8,
+    deepest_level: u8,
+    tile_size: u32,
+    bounding_rect: proto::Rect,
+}
+
+// Checks wheter all the elements in the iterator have the same value and returns it.
+// There must be at least one element in the iterator.
+fn check_all_the_same<I, V>(mut iterator: I) -> Option<V>
+where
+    I: Iterator<Item = V>,
+    V: std::cmp::PartialEq,
+{
+    let first = iterator.next().expect("Iterator cannot be empty");
+    if iterator.all(|element| element == first) {
+        Some(first)
+    } else {
+        None
+    }
+}
+
+fn validate_metadata(metadata: &Vec<proto::Meta>) -> Metadata {
+    assert!(!metadata.is_empty(), "No meta.pb files found.");
+    let (somes, nones): (Vec<Option<NodeId>>, Vec<Option<NodeId>>) = get_root_nodes(metadata)
+        .iter()
+        .partition(|res| res.is_some());
+    assert!(!somes.is_empty(), "All quadtrees are empty.");
+    if !nones.is_empty() {
+        println!(
+            "{} out of {} quadtrees are empty.",
+            nones.len(),
+            metadata.len()
+        );
+    }
+    // The unwrap below is is safe.
+    let root_nodes: FnvHashSet<NodeId> = somes.iter().map(|node_id| node_id.unwrap()).collect();
+    assert_eq!(root_nodes.len(), somes.len(), "Not all roots are unique.");
+    let level = check_all_the_same(root_nodes.iter().map(|node| node.level()))
+        .expect("Not all roots have the same level.");
+    let version = check_all_the_same(metadata.iter().map(|meta| meta.get_version()))
+        .expect("Not all meta files have the same version.");
+    assert_eq!(version, xray::CURRENT_VERSION);
+    let deepest_level = check_all_the_same(metadata.iter().map(|meta| meta.get_deepest_level()))
+        .expect("Not all meta files have the same deepest level.") as u8;
+    let tile_size = check_all_the_same(metadata.iter().map(|meta| meta.get_tile_size()))
+        .expect("Not all meta files have the same tile size.");
+    // TODO: check whether all root nodes have the same bound_rect.
+    let bounding_rect = metadata[0].get_bounding_rect().clone();
+
+    let mut nodes = FnvHashSet::default();
+    for meta in metadata {
+        nodes.extend(meta.get_nodes().iter().map(|proto| NodeId::from(proto)));
+    }
+
+    Metadata {
+        root_nodes,
+        level,
+        deepest_level,
+        tile_size,
+        bounding_rect,
+        nodes,
+    }
+}
+
+fn write_metadata(metadata: Metadata, output_directory: &Path) {
```

Comment:
How about having a `From` implementation for `Meta` instead? Then we could also reuse it for replacing the meta generation code at the end of `build_xray_quadtree` in `xray/src/generation.rs`.

Reply:
I like this idea

Reply:
Since there exists `Meta::from_proto`, for symmetry I added `Meta::to_proto`